google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java

```diff
@@ -1296,15 +1296,12 @@ TransactionSelector getTransactionSelector() {
 
     @Override
     public void onError(SpannerException e) {
-      if (e.getErrorCode() == ErrorCode.ABORTED) {
-        long delay = -1L;
-        if (e instanceof AbortedException) {
-          delay = ((AbortedException) e).getRetryDelayInMillis();
-        }
+      long delay = SpannerException.extractRetryDelay(e.getCause());
+      if (e.getErrorCode() == ErrorCode.ABORTED || (e.getErrorCode() ==
```

Comment:
I don't think adding this condition here is correct since this is used to decide whether to retry the whole transaction (and not just the individual rpc). But RESOURCE_EXHAUSTED should behave like UNAVAILABLE and only retry that specific rpc. You already made the change in `isRetryable` in `SpannerExceptionFactory` which will achieve that effect except that you need to also use the delay specified.